### PR TITLE
Deprecate Cluster Client #27242

### PR DESCRIPTION
**Pull request by patriknw**
Refs #27242

See also sample in https://github.com/akka/akka-samples/pull/120

### CODE REVIEW

File: akka-cluster-tools/src/main/scala/akka/cluster/client/ClusterClient.scala [hunk 1]
```diff
@@ -271,15 +275,18 @@ final case class ContactPoints(contactPoints: Set[ActorPath]) {
     contactPoints.asJava
 }
 
+@deprecated("Use Akka gRPC instead.", since = "2.6.0")
 object ClusterClient {
 
   /**
    * Scala API: Factory method for `ClusterClient` [[akka.actor.Props]].
    */
+  @deprecated("Use Akka gRPC instead.", since = "2.6.0")
```
Comment (johanandren) [inline akka-cluster-tools/src/main/scala/akka/cluster/client/ClusterClient.scala]: Put the url of migrate-to-akkka-grpc in the docs in the deprecation message for easy finding it?

File: akka-docs/src/main/paradox/cluster-client.md [hunk 2]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
 contacts can be fetched and a new cluster client started.
+
+## Migration to Akka gRPC
+
+Cluster Client is deprecated and it is not advised to build new applications with Cluster Client.
+We recommend using [Akka gRPC](https://doc.akka.io/docs/akka-grpc/current/index.html) over using Cluster Client.
+
+Cluster Client is convenient to use if you have Akka actors on both client and server side, but
+a more decoupled, and therefore better, solution would be to use Akka gRPC and define
+application specific protocol buffer messages and gRPC service calls. The benefits would be:
```
Comment (johanandren) [inline akka-docs/src/main/paradox/cluster-client.md]: "The benefits are:"

File: akka-docs/src/main/paradox/cluster-client.md [hunk 3]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
+The `ClusterClient` actor delegates those messages to the gRPC client and on the
+server side those are translated and delegated to the destination actors that
+are registered via the `ClusterClientReceptionist` in the same way as in the original.
+
+The application specific messages are wrapped and serialized with Akka serialization.
+
+### Differences
+
+Apart from that the underlying implementation is using gRPC instead of Actor messages
+and Akka Remoting it's worth pointing out the following differences.
```
Comment (johanandren) [inline akka-docs/src/main/paradox/cluster-client.md]: Maybe make it more clear that the differences below is about the migration-step above and not about Akka gRPC in general?

"... it's worth pointing out the following differences with this Cluster Client emulation style Akka gRPC usage." 

or something along those lines.

**Review by johanandren — APPROVED** [review]
LGTM with some small potential improvements.

File: akka-docs/src/main/paradox/cluster-client.md [hunk 4]
```diff
@@ -1,14 +1,19 @@
 # Classic Cluster Client
 
+@@@ warning
+
+Cluster Client is deprecated in favor of using [Akka gRPC](https://doc.akka.io/docs/akka-grpc/current/index.html). 
+It is not advised to build new applications with Cluster Client, and existing users @ref[should migrate](#migration-to-akka-grpc). 
+
+@@@
+
 @@@ note
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]: After this and https://github.com/akka/akka/pull/27617 merge I'll swing back here and update this below to the new includes.

File: akka-docs/src/main/paradox/cluster-client.md [hunk 5]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
 are entirely dynamic and the entire cluster might shut down or crash, be restarted on new addresses. Since the
 client will be stopped in that case a monitoring actor can watch it and upon `Terminate` a new set of initial
 contacts can be fetched and a new cluster client started.
+
+## Migration to Akka gRPC
+
+Cluster Client is deprecated and it is not advised to build new applications with Cluster Client.
+We recommend using [Akka gRPC](https://doc.akka.io/docs/akka-grpc/current/index.html) over using Cluster Client.
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]: Duplicate of the warning on line 3

File: akka-docs/src/main/paradox/cluster-client.md [hunk 6]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
+
+Cluster Client is convenient to use if you have Akka actors on both client and server side, but
+a more decoupled, and therefore better, solution would be to use Akka gRPC and define
+application specific protocol buffer messages and gRPC service calls. The benefits would be:
+
+* Easier to update clients and servers independent of each other.
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]: Nit: bullet points don't need a period at the end

File: akka-docs/src/main/paradox/cluster-client.md [hunk 7]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
+* More well defined protocol between client and server.
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]:
```suggestion
* Improved protocol definition between client and server
```

File: akka-docs/src/main/paradox/cluster-client.md (hunk 5 above)
Comment (patriknw) [reply akka-docs/src/main/paradox/cluster-client.md]: yes, but this is at the end of the page so deserves repeating

File: akka-docs/src/main/paradox/cluster-client.md [hunk 8]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
+* More secure solution by using TLS for gRPC (HTTP/2) instead of exposing Akka Remoting outside the Akka Cluster.
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]: I'd bump this to the top bullet

Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]:
```suggestion
* Improved security by using TLS for gRPC (HTTP/2) versus exposing Akka Remoting outside the Akka Cluster
```

File: akka-docs/src/main/paradox/cluster-client.md [hunk 9]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
+* The clients don't even have to be using Akka.
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]:
```suggestion
* Clients do not need to use Akka
```

File: akka-docs/src/main/paradox/cluster-client.md (hunk 6 above)
Comment (patriknw) [reply akka-docs/src/main/paradox/cluster-client.md]: curious to learn: until one of them becomes two sentences?

File: akka-docs/src/main/paradox/cluster-client.md [hunk 10]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
+* See also [gRPC vs Akka Remoting](https://doc.akka.io/docs/akka-grpc/current/whygrpc.html#grpc-vs-akka-remoting)
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]:
```suggestion
* See also [gRPC versus Akka Remoting](https://doc.akka.io/docs/akka-grpc/current/whygrpc.html#grpc-vs-akka-remoting)
```

File: akka-docs/src/main/paradox/cluster-client.md [hunk 11]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
+
+Existing users of Cluster Client may migrate directly to plain Akka gRPC, or via a migration
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]:
```suggestion
Existing users of Cluster Client may migrate directly to Akka gRPC, or via a migration
```

File: akka-docs/src/main/paradox/cluster-client.md (hunk 8 above)
Comment (patriknw) [reply akka-docs/src/main/paradox/cluster-client.md]: I was also thinking that service discovery could be a pro

File: akka-docs/src/main/paradox/cluster-client.md [hunk 12]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
+step that requires less re-write of the application. That migration step is described here but we recommend
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]:
```suggestion
step that requires less re-write of the application. That migration step is described below, but we recommend
```

File: akka-docs/src/main/paradox/cluster-client.md [hunk 13]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
+against using the approach for new applications. New applications should use plain Akka gRPC.
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]:
```suggestion
new applications use Akka gRPC.
```

File: akka-docs/src/main/paradox/cluster-client.md [hunk 14]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
+
+An example is provided to illustrate an approach to migrate from the deprecated Cluster Client to Akka gRPC,
+with minimal changes to your existing code. The example is intended to be copied and adjusted to your needs.
+It will not be provided as a published artifact.
+
+* [akka-samples/akka-sample-cluster-cluster-client-grpc-scala](https://github.com/akka/akka-samples/tree/2.6/akka-sample-cluster-cluster-client-grpc-scala) implemented in Scala
+* [akka-samples/akka-sample-cluster-cluster-client-grpc-java](https://github.com/akka/akka-samples/tree/2.6/akka-sample-cluster-cluster-client-grpc-java) implemented in Java
+
+The example is still using an actor on the client side to have an API that is very close
+to the original Cluster Client. The messages this actor can handle corresponds to the
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]:
```suggestion
to the original Cluster Client. The messages this actor can handle correspond to the
```

File: akka-docs/src/main/paradox/cluster-client.md [hunk 15]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
+@ref:[Distributed Pub Sub](distributed-pub-sub.md) messages on the server side, such as
+`ClusterClient.Send` and `ClusterClient.Publish`.
+
+The `ClusterClient` actor delegates those messages to the gRPC client and on the
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]:
```suggestion
The `ClusterClient` actor delegates those messages to the gRPC client, and on the
```

File: akka-docs/src/main/paradox/cluster-client.md [hunk 16]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
+server side those are translated and delegated to the destination actors that
+are registered via the `ClusterClientReceptionist` in the same way as in the original.
+
+The application specific messages are wrapped and serialized with Akka serialization.
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]: Add the link to akka serde 
```suggestion
The application specific messages are wrapped and serialized with Akka Serialization.
```

File: akka-docs/src/main/paradox/cluster-client.md [hunk 17]
```diff
@@ -227,3 +232,62 @@ This can be useful when initial contacts are provided from some kind of service
...
+
+### Differences
+
+Apart from that the underlying implementation is using gRPC instead of Actor messages
```
Comment (helena) [inline akka-docs/src/main/paradox/cluster-client.md]:
```suggestion
Aside from the underlying implementation using gRPC instead of Actor messages
```

**Comment by helena** [comment]
Looks good, I just suggested a few sentence restructures.

File: akka-docs/src/main/paradox/cluster-client.md (hunk 6 above)
Comment (helena) [reply akka-docs/src/main/paradox/cluster-client.md]: I remember it being that way for scientific writing though I dug into the minutia for technical writing from Oxford and Cambridge U, Google, Microsoft. I think it's fine if we don't go with those rules. What you have is ok.

File: akka-docs/src/main/paradox/cluster-client.md (hunk 8 above)
Comment (helena) [reply akka-docs/src/main/paradox/cluster-client.md]: Definitely a pro.

**Review by helena — APPROVED** [review]
LGTM

**Pull request merged**